### PR TITLE
feat: pebbleds profile and plugin

### LINKED ISSUE #10347: Support pebble
## TODO

- [x] include https://github.com/ipfs/go-ds-pebble as built-in plugin (similar to badger atm)
- [x] update https://github.com/ipfs/kubo/blob/master/docs/config.md#profiles to include pebble 
  - [ ] (TBD, sounds like) do we want to allow flatfs (for blocks)+pebble for pins? if so,  we need separate profile for applying it only as leveldb replacement, and separate one for `/blocks` (blockstore) 
- [x] update https://github.com/ipfs/kubo/blob/master/docs/datastores.md to include pebble 
- [x] ~(TBD, probably a nice to have but not required) add support to https://github.com/ipfs/ipfs-ds-convert to allow conversion of existing repos?~ 
   - we've  decided to not write conversion tools, in practice to switch datastore type, people created new cluster node and transferred data to it, as it was safer and also performed GC of unpinned data for free.
- [x] find a way to deal  with ratcheting updates failing (https://github.com/ipfs/kubo/issues/10347#issuecomment-2457485652) See: #10789 
  - [x] expose `FormatMajorVersion` and write docs in [docs/datastores.md#pebbleds](https://github.com/ipfs/kubo/blob/master/docs/datastores.md#pebbleds) ?
- [x] add warning to users running badger, nudging them to switch to flatfs or pebble 
- [ ] replace leveldb with pebble in the default settings
- [ ] soft badger deprecation
   - [ ] remove [badgerv1 profile](https://github.com/ipfs/kubo/blob/master/docs/config.md#badgerds-profile), so new users won't be able to use it, unless they had preexisting config
   - [ ] detect when badgerv1 is used and refuse to start unless `KUBO_ALLOW_DEPRECATED_BADGERV1=true` is set in env – creates friction, forcing people to notice

## Summary

Include pebble as built-in plugin.
It provides meaningful alternative to leveldb as the datastore, and may be better than badger1 as well.

## Background

It is 2024, and we still only have flatfs, leveldb and necromancy-level badgerv1 (!) as datastore options. 

We got positive feedback about pebble, some examples:

https://ipfscluster.io/documentation/guides/datastore/

> 
> [Pebble](https://github.com/cockroachdb/pebble) is a high performant
> backend from Cochroachdb, used by default in Cluster:
> 
> * Proven to work well on very large pinsets.
> * Best disk-usage compared to the rest. No need to trigger GC cycles for space reclaim.
> * Performance and memory usage seems on par with Badger3, and behaves better than Badger on both counts.
> * Behaves correctly with default settings but we bump them up a bit.
> * 0-delay startup times, even with very large amounts of data.
> * Options support compression (we chose to leave it enabled by default).
> * The Pebble project is officially alive and maintained.
> * Pebble only runs on 64-bit architectures.
> * One key difference with Badger3 is that Pebble stores keys and values
>   together and any lookup for a key will also read the values, while Badger3
>   can store keys and values separately (i.e. keys only in the index, which can
>   be loaded onto memory when small enough).

https://github.com/ipfs/go-ds-pebble/issues/29:

> After changing leveldb store to pebble store, the speed of GC has increased by at least dozens of times. So it's not flatfs that's to blame for slow GC, it's leveldb. I also tried leveldb and pebble as blockstore, but the CPU and memory usage is unacceptable.

Right now, to use go-ds-pebble one needs to build external plugin.

## Proposed change

Include https://github.com/ipfs/go-ds-pebble in standard kubo build, just like we do with legacy badger1.

This will
- provide meaningful alternative to leveldb and badgerd
- allow community to use it without having to build external plugin and report back

### PR DESCRIPTION
**Pull request by gammazero**
Pebble provides a high-performance alternative to leveldb as the datastore, and will serve as a replacement for badger1.

There are a number of tuning parameters available for tuning pebble's performance to your specific needs. Default values are used for any parameters that are not configured or are set to their zero-value. 

Requires https://github.com/ipfs/go-ds-pebble/pull/39

Closes #10347

### CODE REVIEW

File: config/init.go [hunk 1]
```diff
@@ -138,6 +138,28 @@ func DefaultDatastoreConfig() Datastore {
...
 
+func pebbleSpec() map[string]interface{} {
+	return map[string]interface{}{
+		"type":   "measure",
+		"prefix": "pebble.datastore",
+		"child": map[string]interface{}{
+			"type":                        "pebbleds",
+			"path":                        "pebbleds",
+			"bytesPerSync":                0,
+			"bisableWAL":                  false,
```
Comment (lidel) [inline config/init.go:149]: typo?

```suggestion
			"disableWAL":                  false,
```

File: docs/config.md [hunk 2]
```diff
@@ -2414,7 +2410,30 @@ You should use this datastore if:
...
+Configures the node to use the pebble high-performance datastore.
+
+Pebble is a LevelDB/RocksDB inspired key-value store focused on performance and internal usage by CockroachDB.
+You should use this datastore if:
+
+- You need a datastore that is focused on performance.
+- You need reliability by default, but may choose to disable WAL for maximum performance when reliability is not critical.
+- This datastore is good for multi-terrabyte data sets.
+- May benefit from tuning depeending on read/write patterns and throughput.
+- Performance is helped significnatly by running on a system with plenty of memory.
```
Comment (lidel) [inline docs/config.md:2430]:
```suggestion
- May benefit from tuning depending on read/write patterns and throughput.
- Performance is helped significantly by running on a system with plenty of memory.
```

File: config/init.go [hunk 3]
```diff
@@ -138,6 +138,28 @@ func DefaultDatastoreConfig() Datastore {
 	}
 }
 
+func pebbleSpec() map[string]interface{} {
```
Comment (lidel) [inline config/init.go:141]: @gammazero I wonder if we would have less maintenance in the future if we simplify things here a bit, and improve  UX:

-  not include config options with hardcoded zero values here, and only have them documented at https://github.com/ipfs/kubo/blob/master/docs/datastores.md
    - this way `ipfs init` does not include them in `config`, we can add new ones in the future, and we are ensuring the user who intents to change default at least had to open the datastores.md docs 
- move pebble configuration options (`pebble.Options`) to `map[string]interface{}` named `pebbleOptions`, and have it empty by default (`{}`) so users have place for adding options, but we don't prefill it with zero vals

**Review by lidel — COMMENTED** [review]
thank you,  pushed some extra docs and e2e test for all three datastore profiles

i think the only remaining topic here is if we should hardcode zero values in `config` –  see the idea inline in `init.go`

File: config/init.go (hunk 3 above)
Comment (gammazero) [reply config/init.go:141]: I completely agree with not including zero values. I removed them from the config profile.

**Review by lidel — APPROVED** [review]
lgtm, let's include it in 0.31.0-rc1 :+1:

**Pull request merged**